feat: ignore some dynamic-value attributes

aldeed opened the pull request:
Resolves #829 

When building cues for an element, attributes that have a value that is either an empty string or appears to be dynamic (machine-generated) are now not used as cues. There are two exceptions: `href` and `src` which would frequently have dynamic values as part of the URL. Other attributes can be made exempt if you can think of reasons, or we could try to make `isDynamic` smarter.

Note that currently a half-dynamic value such as `input-d8sdgjk3` will be used. This is up for debate.

**Update**

Above is the original description, but here's how this actually works at the point of merging it.

- I left `isDynamic` pretty much the same
- But instead of running attribute values through `isDynamic`, I created a new function `getValueMatches` that attempts to find all the static portions of the value string that we can match on
- And then it will use the appropriate CSS selector operator based on what it finds (starts with, ends with, contains, or equals)
- And it still skips this for all `DYNAMIC_VALUE_OK` attributes, and now also for all preferred attributes.
- Also, rather than trying to specifically list HTML tags in the `words` list, I added the `html-tags` package and merged in the array provided by that package, which fixed some existing tests that were otherwise failing

So for example you now get selectors like this:

```js
await page.click('[name="nonDynamicInput"]'); // name="nonDynamicInput"
await page.click('[name^="input-"]'); // name="input-bu32879fDi"
await page.click('[name$="-check"]'); // name="bu32879fDi-check"
await page.click('[name*="-blue-"]'); // name="f89ndrn4-blue-5"
```

jperl commented on src/web/cues.ts:
```diff
@@ -2,6 +2,8 @@ import { getAttribute } from './attribute';
 import { getElementText } from './selectorEngine';
 import { isDynamic } from './isDynamic';
 
+const DYNAMIC_VALUE_OK_ATTRIBUTES = ['href', 'src'];
```
We should include "value" too.

jperl reviewed (approved):
Just need to fix the selector tests then 👍

jperl commented:
@aldeed Seems like you were able to "just" fix it 😜. Love your approach. Everything looks good, at a quick glance I find `getValueMatchSelector` hard to understand.

aldeed commented:
Yes, it's very hard to understand. I wrote it a simpler way first, but I think it would have been pretty inefficient. I'll add more comments and think about whether anything could be simplified.

aldeed commented:
@jperl I rewrote the new function to return multiple matches, which then become multiple attribute value cues. I realized this is better because then cue optimization logic can pick what's best. This had the added benefit of making the code easier to understand, but I also added a bunch of comments explaining the flow.

Sandbox needs to be published, but 🤞 this will be all good now.

jperl commented:
@aldeed The refactor & comments are great 💯 I was able to follow it. Publishing sandbox now.

jperl commented:
There is just one remaining issue, we run into selectors with [multiple split characters](https://github.com/qawolf/qawolf/blob/main/test/web/isDynamic.test.ts#L44). I updated [this test](https://github.com/qawolf/qawolf/pull/834/commits/f0e573749d37a81e12e71c5dbe3628cf569b23cf) and it shows there is a current bug with that.

I think the behavior we want is that if there are multiple, the `ValueMatch` will only use the one right before/after.

aldeed commented:
I fixed that issue, but I made it include all split characters rather than just the immediate ones, partially because that was easiest and partially because it seems to me that more characters would be a more likely match? I don't have strong feelings about that, so I can easily keep just the first before/after if you're sure that's best.

jperl commented on src/web/isDynamic.ts:
```diff
@@ -182,12 +182,14 @@ export const getValueMatches = (
+      let backwardPosition = currentPosition - 1;
+      let previousCharacter = value[backwardPosition];
+      while (SPLIT_CHARACTERS.includes(previousCharacter)) {
+        if (tokenType === 'static') currentStaticBlock = previousCharacter + currentStaticBlock;
```
Is this if statement necessary? The tokenType should be static https://github.com/qawolf/qawolf/pull/834/commits/3c94c7f510957ca5d1856f7b59ee6f19230ffcc7#diff-457f25ee3f7d5025b30b8dfb3d70da85R183

aldeed replied:
Good catch. Fixed.

jperl commented:
Awesome work. Merge it! 🚢

jperl commented on src/web/cues.ts:
```diff
@@ -1,6 +1,8 @@
-import { isDynamic } from './isDynamic';
+import { getValueMatches, isDynamic } from './isDynamic';
+
+const DYNAMIC_VALUE_OK_ATTRIBUTES = ['href', 'src', 'value'];
```
let's add placeholder to this

aldeed replied:
Done

jperl commented:
2 small comments. Then merge it 🚢 . Awesome work 🎉
